# Using a pre-built CNN

In this notebook we are going to use a CNN model that was trained on the ImageNet dataset. The dataset consists of 1.4 million images in 1,000 categories. There are a number of models trained on this dataset. We will use one calle VGG16, which was developed by Karen Simonyan and Andrew Zisserman in 2014.

## Let's jump in.

First the preliminaries:

In [1]:
import keras
keras.__version__

Using TensorFlow backend.


'2.1.3'

Now let's import the VGG16 model:

In [4]:
from keras.applications import VGG16

vgg = VGG16(weights='imagenet',
                  input_shape=(224, 224, 3))

Let's take a look at the structure of the model:

In [5]:
vgg.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         (None, 224, 224, 3)       0         
_________________________________________________________________
block1_conv1 (Conv2D)        (None, 224, 224, 64)      1792      
_________________________________________________________________
block1_conv2 (Conv2D)        (None, 224, 224, 64)      36928     
_________________________________________________________________
block1_pool (MaxPooling2D)   (None, 112, 112, 64)      0         
_________________________________________________________________
block2_conv1 (Conv2D)        (None, 112, 112, 128)     73856     
_________________________________________________________________
block2_conv2 (Conv2D)        (None, 112, 112, 128)     147584    
_________________________________________________________________
block2_pool (MaxPooling2D)   (None, 56, 56, 128)       0         
__________

## Question: How many convolutional layers are there in VGG16?

# Classify a photo
Let's try to classify this photo:

![](http://zacharski.org/files/courses/cs370/vggTest/poodle2.jpg)

(or you can classify one of your own). Go ahead and download the photo.

then set the filename to match the location of your downloaded photo:


In [26]:
filename = '/Users/raz/data/vggTest/poodle2.jpg'

## Preprocess the image


In [27]:
from keras.preprocessing.image import load_img
from keras.preprocessing.image import img_to_array

image = load_img(filename, target_size=(224, 224))

# now convert it to a numpy array
image = img_to_array(image)

# reshape data for the model
image = image.reshape((1, image.shape[0], image.shape[1], image.shape[2]))

Finally we need to preprocess that numpy representation of the image so it can ber accepted by VGG:

In [31]:
from keras.applications.vgg16 import preprocess_input
# prepare the image for the VGG model
image = preprocess_input(image)

This seems like a lot of steps, and I keep thinking there must be a better preprocessing image pipeline but I don't see it.

## Finally Classification

In [32]:
# predict the probability across all output classes
predictions = vgg.predict(image)

In [29]:
from keras.applications.vgg16 import decode_predictions
# convert the probabilities to class labels
labels = decode_predictions(predictions, top=3)
labels = labels[0]
for label in labels:
    print('%s (%.2f%%)' % (label[1], label[2]*100))

standard_poodle (94.32%)
Irish_water_spaniel (3.83%)
miniature_poodle (1.48%)


## Another picture
What will it do with this picture?

![](http://zacharski.org/files/courses/cs370/vggTest/borderCollie.jpg)


In [25]:
dogPic = '/Users/raz/data/vggTest/borderCollie.jpg'

In [33]:
from keras.preprocessing.image import load_img
from keras.preprocessing.image import img_to_array

image = load_img(dogPic, target_size=(224, 224))

# now convert it to a numpy array
image = img_to_array(image)

# reshape data for the model
image = image.reshape((1, image.shape[0], image.shape[1], image.shape[2]))

from keras.applications.vgg16 import preprocess_input
# prepare the image for the VGG model
image = preprocess_input(image)

In [34]:
# predict the probability across all output classes
predictions = vgg.predict(image)

from keras.applications.vgg16 import decode_predictions
# convert the probabilities to class labels
labels = decode_predictions(predictions, top=3)
labels = labels[0]
for label in labels:
    print('%s (%.2f%%)' % (label[1], label[2]*100))

Border_collie (38.23%)
Cardigan (15.72%)
collie (10.80%)


## The task

Can you create a function `predict(path_to_images)` where `path_to_images` is the path to a directory containing images. 
The output will be one line per image in the directory of the form:

     poodle2.jpg       standard_poodle (94.32%) Irish_water_spaniel (3.83%) miniature_poodle (1.48%)
     borderCollie.jpg  Border_collie (38.23%) Cardigan (15.72%) collie (10.80%)
     
The following Python functions may be helpful.

In [2]:
import os


for dirname, dirnames, filenames in os.walk('/Users/raz/data/vggTest'):
    
    for filename in filenames:
        print(filename, os.path.join(dirname, filename))

  

.DS_Store /Users/raz/data/vggTest/.DS_Store
camping.jpg /Users/raz/data/vggTest/camping.jpg
borderCollie.jpg /Users/raz/data/vggTest/borderCollie.jpg
poodle1.jpg /Users/raz/data/vggTest/poodle1.jpg
poodle2.jpg /Users/raz/data/vggTest/poodle2.jpg


### Then try it on a directory of some photos. 
Here is [a list of the classes the classifier recognizes](https://gist.github.com/yrevar/942d3a0ac09ec9e5eb3a)